src/protocol/boolean/bit_decomposition.rs

```diff
@@ -0,0 +1,210 @@
+        ctx: [SemiHonestContext<'_, F>; 3],
+        record_id: RecordId,
+        a: F,
+    ) -> Result<Vec<F>, Error>
+    where
+        Standard: Distribution<F>,
+    {
+        let [c0, c1, c2] = ctx;
+        let mut rand = StepRng::new(1, 1);
+
+        let s = share(a, &mut rand);
+
+        let result = join3(
+            BitDecomposition::execute(c0, record_id, &s[0]),
+            BitDecomposition::execute(c1, record_id, &s[1]),
+            BitDecomposition::execute(c2, record_id, &s[2]),
+        )
+        .await;
+
+        // bit-decomposed values must have the same bit length of the target field
+        assert_eq!(F::Integer::BITS as usize, result[0].len());
+        assert_eq!(F::Integer::BITS as usize, result[1].len());
+        assert_eq!(F::Integer::BITS as usize, result[2].len());
+
+        let bits = (0..result[0].len())
+            .map(|i| validate_and_reconstruct(&result[0][i], &result[1][i], &result[2][i]))
+            .collect::<Vec<_>>();
+
+        Ok(bits)
+    }
+
+    #[ignore]
+    #[tokio::test]
+    pub async fn fp31() -> Result<(), Error> {
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<Fp31>(&world);
+        let [c0, c1, c2] = ctx;
+        let mut rng = rand::thread_rng();
+
+        for i in 0..10 {
+            let input = rng.gen::<Fp31>();
+            let result = bit_decomposition(
+                [c0.clone(), c1.clone(), c2.clone()],
+                RecordId::from(i),
+                input,
+            )
+            .await?;
+
+            if input.as_u128() == 0 {
+                // if the protocol's input is 0, the output could either be
+                // a bitwise sharing of 0 or the prime (in the integers).
+                let x = bits_to_value(&result);
+                assert!(x == 0 || x == u128::from(Fp31::PRIME));
+            } else {
+                // otherwise, the reconstructed integer (not the field) must
+                // be in the range of the field `0..p`.
+                assert_eq!(input.as_u128(), bits_to_value(&result));
+            }
+        }
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    pub async fn fp32_bit_prime() -> Result<(), Error> {
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<Fp32BitPrime>(&world);
+        let [c0, c1, c2] = ctx;
+        let mut rng = rand::thread_rng();
+
+        for i in 0..2 {
+            let input = rng.gen::<Fp32BitPrime>();
+            let result = bit_decomposition(
+                [c0.clone(), c1.clone(), c2.clone()],
+                RecordId::from(i),
+                input,
+            )
+            .await?;
+
+            if input.as_u128() == 0 {
+                // if the protocol's input is 0, the output could either be
+                // a bitwise sharing of 0 or the prime (in the integers).
+                let x = bits_to_value(&result);
+                assert!(x == 0 || x == u128::from(Fp32BitPrime::PRIME));
```

Comment:
Should we just test the zero case specifically? Also, I remain unconvinced that it should yield a bitwise decomposition of the prime in the case the input is zero.